### PR TITLE
Qt Location based map

### PR DESCRIPTION
**Pull request by neolit123**
This is a side-by-side implementation of the Qt Location based map next to our already existing Marble map. Both map implementations now now exist in the codebase  and can be swapped.

It's a bit of patch bomb with ~120 commits and while it goes back and forth here and there, the patches try to remain small and follow my thinking process...still, if some refactoring is needed, let me know.

**How it works:**

The best thing about this map widget is that it can be used _for both_ the mobile and desktop versions. The mobile version support is not implemented, as it needs some decision making. The desktop version support begins with the class `dekstop-widgets/mapwidget.cpp` which is a `QQuickWidget` that loads
a QML file `mobile-widgets/qml/MapWidget.qml`. All the NO_MARBLE abstraction is taken care of and MapWidget methods are used instead of Globe methods. The code in mapwidget.cpp code is minimal as this class desktop-version specific. The magic happens in the class `mobile-widgets/qmlmapwidgetheloper.cpp` which is the backend of the map widget. It is instantiated in `MapWidget.qml` and both the mobile and desktop version need to communicated with it. A new model is added - `qt-models/maplocationmodel.cpp` that deals with the population of markers on the map.

The map itself supports a context menu (`mobile-widgets/qml/MapWidgetContextMenu.qml`) with easy support for any number of options. Markers on this map can be edited by dragging instead of a double-click. The right mouse button is not used, while double-click is used to zoom in over a location or a marker. This is easy to translate to the mobile version without abstraction, plus on mobile the Map QML object should already support functionality like pinch zoom.

For the mobile version the infrastructure should be already in place. the MapWidget needs to instantiated somewhere, some extra signal/slot handling needs to be added and it should be good to go (hopefully).

**How to build:**
- you need Qt 5.9 (i haven't tested with newer or on anything but Win32)
- build using `cmake -DNO_MARBLE=1`

**Next steps would be:**
- receive code reviews
- receive feedback on the functionality
- remove Marble from the codebase

### CODE REVIEW

File: desktop-widgets/CMakeLists.txt [hunk 1]
```diff
@@ -70,6 +70,7 @@ set(SUBSURFACE_INTERFACE
 	kmessagewidget.cpp
 	mainwindow.cpp
 	mapwidget.cpp
+	../mobile-widgets/qmlmapwidgethelper.cpp
```
Comment (dirkhh) [inline desktop-widgets/CMakeLists.txt]: yeah, that's REALLY ugly - as is the overflowing commit message.
not enough to ask you to redo this massive pull request, but just saying...

File: desktop-widgets/mapwidget.cpp [hunk 2]
```diff
@@ -6,11 +6,14 @@
 #include "mapwidget.h"
```
Comment (dirkhh) [inline desktop-widgets/mapwidget.cpp]: (I can't comment on the commit message, so I comment here)
That reference of the "previous commit" appears to reference something that isn't there anymore...

Comment (neolit123) [reply desktop-widgets/mapwidget.cpp]: which commit hash?

File: desktop-widgets/CMakeLists.txt (hunk 1 above)
Comment (neolit123) [reply desktop-widgets/CMakeLists.txt]: i will fix the commit message overflow.

File: desktop-widgets/mapwidget.cpp (hunk 2 above)
Comment (dirkhh) [reply desktop-widgets/mapwidget.cpp]: aww that's bad - poor GitHub UI. It's bc8f5c7401c1d417f4668bd42f01a50169d3d143 - but the comment shows up on every commit to that file... sorry about that.

File: mobile-widgets/qml/MapWidget.qml [hunk 3]
```diff
@@ -69,17 +69,10 @@ Item {
 		ParallelAnimation {
 			id: mapAnimation
 			CoordinateAnimation {
-				target: map
-				property: "center"
-				to: map.newCenter
-				duration: 2000
+				target: map; property: "center"; to: map.newCenter; duration: 2000;
```
Comment (dirkhh) [inline mobile-widgets/qml/MapWidget.qml]: 200a6bf5047bc988f7a4be2d380725e3c1d20a36: we are not 100% consistent with that, but in general we went with one property per line unless there is a reason not to...

File: mobile-widgets/qml/MapWidget.qml [hunk 4]
```diff
@@ -32,6 +32,8 @@ Item {
 			activeMapType = mapType.SATELLITE
 		}
 
+		onZoomLevelChanged: mapHelper.calculateSmallCircleRadius(map.center)
```
Comment (dirkhh) [inline mobile-widgets/qml/MapWidget.qml]: 68ed3e0c8c62574fb8247f923554469bb007d384: another commit with overflowing commit message

File: mobile-widgets/qml/MapWidget.qml (hunk 3 above)
Comment (neolit123) [reply mobile-widgets/qml/MapWidget.qml]: it one of the few last patches i make most animations to be on single lines, so maybe i need to expand all properties to be on separate lines everywhere.

Comment (dirkhh) [reply mobile-widgets/qml/MapWidget.qml]: I'll be honest - I don't feel THAT strongly about it. It was mainly a comment when I saw the explicit change... this makes the code more compact
The question is "are these arguments to a function call or are these statements in a procedure". It's somewhere in between, isn't it?
I'm ok to leave them in the more dense format

File: mobile-widgets/qmlmapwidgethelper.cpp [hunk 5]
```diff
@@ -59,7 +59,22 @@ void MapWidgetHelper::reloadMapLocations()
 
 void MapWidgetHelper::selectedLocationChanged(MapLocation *location)
 {
-	qDebug() << location;
+	int idx;
```
Comment (dirkhh) [inline mobile-widgets/qmlmapwidgethelper.cpp]: 38af4120f7139fede5ccdb8272d382895ff46c35 : and another overflowing commit message

**Comment by dirkhh** [comment]
OK, I've done a fairly coarse review. Miniscule changes, but by and large I think this should go in so it gets actual testing and feedback... since there were some changes that you were fixing, I'll wait with the actual pull until you give me the go-ahead

**Comment by neolit123** [comment]
thanks for the review. i will make the update to the pull request during the day and notify about it.

**Comment by neolit123** [comment]
done with the update.

**Pull request merged**